Include Google Scholar's required aggregation pages out-of-the-box

noahwsmith opened the pull request:
(my first Islandora PR, so please let me know if this needs improvement/cleanup in any way...)


**JIRA Ticket**: 1816 - this delivers functionality expected from recent conversations with the Google Scholar team (see below).

* Other Relevant Links (Google Groups discussion, related pull requests, Release pull requests, etc.)

# What does this Pull Request do?

The purpose of this is to ensure that automatic crawlers are able to discover and fetch the URLs of all articles, as well as to periodically refresh their content. In Google Scholar setup (https://partnerdash.google.com/partnerdash/d/scholarinclusions#p:id=new), you are asked to provide the following:

Please provide the URL of a page listing all years available.
Please provide examples of URLs to pages that provide a list of articles for a given year.
Please provide examples of individual abstract pages.
Please provide examples of full-text article URLs. Please pick some recently added articles as well as some articles that were added a while ago.
The first two will require creating custom displays, while the third and fourth require no additional functionality. In addition, for larger repositories, GS recommends providing a page that lists items added in the last two weeks. So this part of the task requires building three pages:

A page that lists years. I think the way to do this is to have a custom page which runs a solr query returning zero rows, and faceted with a facet.date.gap of a year.
A page listing articles, filtered by a year parameter in the url.
A page showing articles added/updated in the past two weeks.


# What's new?

These requirements are intended to be met in this commit as follows:

- New pages added:
  - /gs_years: Lists all years when articles were published.
  - /gs_year/%: Lists all articles published in a given year
  - /gs_updated: Lists all articles added or updated in the repository in the past two weeks
  - /gs_abstract/%: Displays a very basic page showing the title and abstract for a given object
- Expanded islandora_scholar configuration form:
  - Added fields to the Google Scholar section for which solr fields to use for publication date and abstract, and which content models to exclude from these pages.
- Documented these enhancements in the islandora_google_scholar README file.

# How should this be tested?

Review new admin settings on `/admin/islandora/solution_pack_config/scholar`, activated by enabling "Render Google Scholar Search Link". Review new Google Scholar submission requirement pages at `gs_years`, `gs_year/:year:`, `gs_abstract/:pid:`. 


# Additional Notes:
This was part of the LASIR project, developed by Born-Digital (Noah Smith and Pat Dunlavey) in consultation with Bryan Brown and Donald Moses. 

# Interested parties
@bryjbrown @dmoses @patdunlavey

Commit:
Requirements:

The purpose of this is to ensure that automatic crawlers are able to discover and fetch the URLs of all articles, as well as to periodically refresh their content. In Google Scholar setup (https://partnerdash.google.com/partnerdash/d/scholarinclusions#p:id=new), you are asked to provide the following:

Please provide the URL of a page listing all years available.
Please provide examples of URLs to pages that provide a list of articles for a given year.
Please provide examples of individual abstract pages.
Please provide examples of full-text article URLs. Please pick some recently added articles as well as some articles that were added a while ago.
The first two will require creating custom displays, while the third and fourth require no additional functionality. In addition, for larger repositories, GS recommends providing a page that lists items added in the last two weeks. So this part of the task requires building three pages:

A page that lists years. I think the way to do this is to have a custom page which runs a solr query returning zero rows, and faceted with a facet.date.gap of a year.
A page listing articles, filtered by a year parameter in the url.
A page showing articles added/updated in the past two weeks.

These requirements are intended to be met in this commit as follows:

- New pages added:
  - /gs_years: Lists all years when articles were published.
  - /gs_year/%: Lists all articles published in a given year
  - /gs_updated: Lists all articles added or updated in the repository in the past two weeks
  - /gs_abstract/%: Displays a very basic page showing the title and abstract for a given object
- Expanded islandora_scholar configuration form:
  - Added fields to the Google Scholar section for which solr fields to use for publication date and abstract, and which content models to exclude from these pages.
- Documented these enhancements in the islandora_google_scholar README file.

noahwsmith commented:
cc @DonRichards and @whikloj too, sorry for leaving you off the original

jonathangreen commented:
@noahwsmith It looks like CI is failing on some [code sniffer issues](https://travis-ci.org/Islandora/islandora_scholar/jobs/560060530#L1429-L1445): 

```
sites/all/modules/islandora_scholar/includes/page_callbacks.inc:
 +27: [normal] Control statements should have one space between the control keyword and opening parenthesis
 +29: [normal] Control statements should have one space between the control keyword and opening parenthesis
 +74: [normal] The $text argument to l http://api.drupal.org/api/function/l/() should be enclosed within t http://api.drupal.org/api/function/t/() so that it is translatable.
 +98: [normal] The first parameter to t() should be a literal string. There should be no variables, concatenation, constants or other non-literal strings there. Read the documentation.
 +100: [normal] The $text argument to l http://api.drupal.org/api/function/l/() should be enclosed within t http://api.drupal.org/api/function/t/() so that it is translatable.
 +116: [normal] Control statements should have one space between the control keyword and opening parenthesis
 +118: [normal] Control statements should have one space between the control keyword and opening parenthesis
 +132: [normal] The $text argument to l http://api.drupal.org/api/function/l/() should be enclosed within t http://api.drupal.org/api/function/t/() so that it is translatable.
 +141: [normal] The $text argument to l http://api.drupal.org/api/function/l/() should be enclosed within t http://api.drupal.org/api/function/t/() so that it is translatable.
 +160: [normal] missing space after comma
 +164: [normal] The $text argument to l http://api.drupal.org/api/function/l/() should be enclosed within t http://api.drupal.org/api/function/t/() so that it is translatable.
 +175: [normal] Control statements should have one space between the control keyword and opening parenthesis
 +177: [normal] Control statements should have one space between the control keyword and opening parenthesis
 +191: [normal] The $text argument to l http://api.drupal.org/api/function/l/() should be enclosed within t http://api.drupal.org/api/function/t/() so that it is translatable.
 +222: [normal] The $text argument to l http://api.drupal.org/api/function/l/() should be enclosed within t http://api.drupal.org/api/function/t/() so that it is translatable.
 +247: [normal] The $text argument to l http://api.drupal.org/api/function/l/() should be enclosed within t http://api.drupal.org/api/function/t/() so that it is translatable.
 +252: [normal] The $text argument to l http://api.drupal.org/api/function/l/() should be enclosed within t http://api.drupal.org/api/function/t/() so that it is translatable.
```

DonRichards commented:
### 2 things I've noticed so far that might need attention.
![Screenshot](https://user-images.githubusercontent.com/2738244/61397970-06db2a80-a899-11e9-8945-9273cae8a4c4.jpg)
This is looking great, but noticed this issue after ingesting 2 items (1 thesis, 1 citation).
> 1st Test : Citation : Date Issued 2019-01-01
> 2nd Test : Thesis : Date Issued 11-23-1969

"1st Test" shows up in both 2019 & 2018 
"2nd Test" does not show on the 2019 or the 1969 page but does show on the latest list. 

The form accepts dates in the following formats YYYY, YYYYMMDD, MMDDYYYY
I think this might be part of the issue here. 

#### 2nd

Something I'm going to suggest, redirecting traffic that navigates to `/gs_year` to automatically default to the current year.

noahwsmith commented:
(thanks- working on these)

Commit:
syntax updates required by sniffer

Commit:
page title translater for sniffer

Commit:
changing year page title to variable before passing to Drupal

noahwsmith commented:
For tomorrow: https://travis-ci.org/Islandora/islandora_scholar/jobs/560276032#L1521

And address Don's findings (I confirm a `gs_years` page bug, and we'll look into a redirect for `gs_year`).

patdunlavey committed:
pat / rebase-lasir-t1-gs_years --- gs_years and gs_year fixes

noahwsmith commented:
(still working on this with @patdunlavey, haven't forgotten y'all)

patdunlavey committed:
pat / rebase-lasir-t1-gs_years --- Previous approach for fixing the years was totally wrong. Turns out that date facets spanning adjacent time gaps, and thereby double-counting, is a known problem, and that the solution is to fudge your start and end dates by a value smaller than the precision of the date indexing. See http://lucene.472066.n3.nabble.com/Date-Facet-duplicate-counts-td525335.html

patdunlavey committed:
pat / rebase-lasir-t1-gs_years --- Refactoring of some code politely requested by Travis CI, and eliminated the redundant gs_year menu item.

Commit:
spacing between control statement and parenthesis

noahwsmith commented:
Ok, that did it on the travis tests.

@DonRichards we fixed the `years` bug - good catch! Unfortunately the `11-23-1969` bug is outside the scope of this module: that date does not generate a `_mdt` solr field because `slurp_all_mods_to_solr` doesn't handle it correctly. All dates that generate a valid solr `mods_originInfo_dateIssued_mdt` field are handled correctly by this PR. 

Solr generates these related fields, from that example:
```
"mods_originInfo_dateIssued_s": "11-23-1969",
        "mods_originInfo_dateIssued_mlt": "11-23-1969",
        "mods_originInfo_dateIssued_t": "11-23-1969",
        "mods_originInfo_dateIssued_ms": [
          "11-23-1969",
          "11-23-1969"
        ],
        "mods_originInfo_dateIssued_mt": [
          "11-23-1969",
          "11-23-1969"
        ],
        "mods_originInfo_dateIssued_ss": "11-23-1969",
```

whereas a valid date entry gives something like 

```
 "mods_originInfo_dateIssued_dt": "2019-01-01T00:00:00Z",
        "mods_originInfo_dateIssued_mt": [
          "2019-01-01T00:00:00.000Z",
          "2019-01-01",
          "2019-01-01T00:00:00.000Z",
          "2019-01-01"
        ],
        "mods_originInfo_dateIssued_mdt": [
          "2019-01-01T00:00:00Z",
          "2019-01-01T00:00:00Z"
        ],
        "mods_originInfo_dateIssued_s": "2019-01-01",
        "mods_originInfo_dateIssued_mlt": "2019-01-01",
        "mods_originInfo_dateIssued_t": "2019-01-01",
        "mods_originInfo_dateIssued_ms": [
          "2019-01-01",
          "2019-01-01"
        ],
        "mods_originInfo_dateIssued_ss": "2019-01-01",
```

noahwsmith commented:
Ready for further review @jonathangreen @whikloj  - does this need to go in today?

whikloj commented:
I don't really have any idea what this does. I looked over the code and couldn't see anything wrong with it, but I could really use a Scholar user to test it. @DonRichards @bryjbrown @dmoses @jonathangreen ??

noahwsmith commented:
The short version is that Google Scholar has some aggregate pages they require for submission, and without this addition, anyone who wants to submit their Islandora Scholar site to Google Scholar has to build them by hand. This addition creates the additional URL patterns and generates the content Google expects as a pre-req for acceptance/indexing.

DonRichards commented:
#### Observation:
I assume the content coming up on the new pages are only content to be delivered to Google Scholar. If that's the case there is a problem.

**/gs_years**: Lists everything except citations
**/gs_year/%**: Lists everything except citations
**/gs_updated**: Lists anything including basic & Large images
**/gs_abstract/%**:  Lists anything including basic & Large images

whikloj commented:
@DonRichards I already looked through this PR but I need someone who had any understanding of Scholar, Google Scholar and (I assume) LASIR to review the functionality. Unfortunately this is an improvement and has now missed code. ;( freeze

patdunlavey commented:
@DonRichards responding to your observations. 

First, a correction: `/gs_abstract/%` is not a list page. It's just a simple object display page (which shows the title and the abstract for a given PID). 

That leaves three cases to consider: `/gs_years`, `/gs_year/%`, and `/gs_updated`.

I double-checked, and there is nothing in this code that makes any assumptions about what is and what is not content or content models that are eligible to be offered to google scholar. Citations have no special status, for example. Ditto for basic and large images. You can tell google scholar to ignore certain content models in on the scholar settings page, but otherwise, everything is fair game.

Our query does incorporate the Solr base filter. Might that be incorrect? Do you have thoughts about that? 

If you enable solr debugging and inspect the solr query, you'll see that the 'fq' section is the same in all three cases (`/gs_years`, `/gs_year/%`, and `/gs_updated`). This is the part of the query which may explicitly filter out content models.

The `/gs_years` and `/gs_year/%` pages can only show objects which have an indexed value for the solr field configured for "Google Scholar Publication Date SOLR Field". This could be a small minority of the content on the site or even no content at all.

The `/gs_updates` page, which shows all added/updated content from the past two weeks, does not require this solr date value, and instead is hard-coded to use `fgs_lastModifiedDate_dt`. So it is potentially much more inclusive for the time period in question. The distinction between when an object was added to the repository, and its publication date, is pretty fundamental. All objects have a `fgs_lastModifiedDate_dt` but it's possible that a small fraction have a publication date. Should we be excluding objects without a publication date from the "Updates" page? That would make it consistent with the `gs_years` and `gs_year/%` results, but could exclude content that we don't want to exclude.

Does anybody else have thoughts about that?

DonRichards commented:
@patdunlavey I assumed the `/gs_abstract/%` would also be filtering the results to only process IR PIDS. 

Thesis/Citation content models provides bibliographic data and references through the object's page meta tags. This module processes scholar material much differently than other content models. In reference to this PR it mainly it adds the the Google Scholar meta-tags from the [MODS values](https://github.com/Islandora/islandora_scholar/blob/7.x/includes/google_scholar.inc). These meta tags confirms to Google Scholar's parsers that the content on the page contains scholarly material and fulfills the [inclusion requirements](https://scholar.google.com/intl/en/scholar/inclusion.html#indexing) except for **2.a.**(_Indexing of content without the meta-tags_).

My big concern here is a statement on their inclusion guidelines, "Incorrect identification of bibliographic data or **references** will lead to poor indexing of your site." --[Inclusion Guidelines for Webmasters](https://scholar.google.com/intl/en/scholar/inclusion.html#indexing). 
Delivering content in a list(references) that isn't scholarly may put us into this misidentification violation.

Maybe a Solr filter like `RELS_EXT_hasModel_uri_s = info:fedora:ir:thesisCModel` would work to isolate the results to just thesis and one for citations. 

I recently ran into a problem with using `fgs_lastModifiedDate_dt` that might be useful to consider. This field is triggered on any modification, like updates, datastream modification etc.. I speculate that even an embargo state change will trigger it. I'd lean towards a published date and maybe  "fgs_createdDate_dt" as a fall back if the publish date is not available.

These are just my 2cents on the topic. I'd still like to hear from others' and their opinions on this. 

@bryjbrown

bryjbrown commented:
With respect to what cmodels to include in the special Google Scholar routes, I would agree that it should be limited to citation/thesis cmodels, not necessarily because thats the most accurate thing to do, but instead because its the most consistent and follows the rest of the pattern that Islandora Scholar has established in saying that thesis/citation cmodels are the only inherently scholarly cmodels. Only citation/thesis cmodels get the special metatags, so I would expect them to be the only types of objects that would be exposed to Google Scholar as well. @DonRichards is correct (as far as I understand) that Google Scholar will penalize you if you misrepresent your content, and I think including things other than citation/thesis cmodel objects fits that description. Many Islandora IRs are actually just part of a shared IR/Digital Archive instance, and without the filtering by cmodel you might accidentally hurt your Google Scholar coverage this way.

As for the use of `fgs_lastModifiedDate_dt`, I don't see an issue with it. Sure, it might include things that didn't seem to Google Scholar to change, but I'd rather have the updated material be too inclusive than use created or published dates that only show new material but not also older material that changed and needs reindexing.

patdunlavey committed:
origin/rebase-lasir-t1 --- addressing issues raised by DonRichards and bryjbrown on the pull request:

- Changed the "excluded content models" configuration to "included content models", and set the default to the citation and thesis content models.
- Some minor tweaks (page titles, no results handling).

https://github.com/Islandora/islandora_scholar/pull/327#issuecomment-514308825
https://github.com/Islandora/islandora_scholar/pull/327#issuecomment-514312995

patdunlavey commented:
@DonRichards I changed the "exclude content models" configuration to "**include** content models", with the defaults being the citation and thesis content models. 

I do think we need to provide the flexibility to catalog other content models, but I could be talked out of providing that configuration in the UI if the consensus is against it. I would note that I have included identifying tags to these solr queries so that one could write a hook_solr_query_alter() to override these filters in any case.

I made a few other small changes to page titles and such to make this more "scholarly".

I really appreciate the effort you guys have taken to review and test this!!

patdunlavey committed:
origin/rebase-lasir-t1 --- removed unintentional change to default date issued solr field

DonRichards commented:
@patdunlavey Might need to lint your code. Restarted Travis just in case it was a misfire.

whikloj commented:
@patdunlavey Just two little spaces needed.
https://travis-ci.org/Islandora/islandora_scholar/jobs/563178862#L995-L997

DonRichards commented:
@noahwsmith would you be able to make those 2 corrections @whikloj suggested?

Commit:
Spacing to satisfy linter

sites/all/modules/islandora_scholar/includes/page_callbacks.inc:
 +37: [normal] Control statements should have one space between the control keyword and opening parenthesis
 +249: [normal] Control statements should have one space between the control keyword and opening parenthesis

noahwsmith commented:
You got it, boss!

bondjimbond commented:
@DonRichards Has this met your requirements for approval?

DonRichards commented:
@bondjimbond It just needs to be tested.

bondjimbond reviewed (approved):
Works as described, and sorely needed! Thanks for this.

DonRichards commented:
20 hours left and counting.

bondjimbond commented:
@noahwsmith I'm getting some weird behaviour in my live site after updating: https://jira.duraspace.org/browse/ISLANDORA-2477

All years from 1 to 9999 are showing up, not just the actually-populated years. Any idea why? https://arcabc.ca/gs_years

noahwsmith commented:
@patdunlavey this seems related to the last material fix to this project - was it intentional to show years with zero content?

bondjimbond commented:
I can't see any reason to show years with zero content.

bondjimbond commented:
@noahwsmith Also I am realizing that they aren't sorted by year, either - they're sorted by number of results.

noahwsmith commented:
Our demo site shows years in order, and without extra rows:
https://lasir-demo.born-digital.com/gs_years (I just pulled the latest code, cleared caches, etc).

Has anyone else updated their `7.x` branch since this was accepted, and could verify either Brandon or my experience? Maybe @dmoses or @DonRichards ?
